Keycloak 26.6.0/26.6.1 exits (code 1) ~100ms after async realm migration completes; migrations not persisted

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

Upgrading Keycloak from 26.5.7 → 26.6.0 (or 26.6.1) on PostgreSQL causes the server to exit with code 1 approximately 100-140ms after realm-level migration completes. No error is logged. The migration_model table never receives a 26.6.x row, so every container restart re-runs the migration of all realms, creating an infinite loop: build → migrate → silently exit → container restart → repeat.                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                    
The migration steps log migrated realm X to 26.6.1 for every realm, and the Liquibase DB lock releases cleanly. Then on the main thread, io.quarkus.runtime.Application prints Stopping application followed by graceful Quarkus shutdown and process exit 1. Downgrading to 26.5.7 (same DB, same env) fixes the problem entirely.                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                    
 **_Version_**                                                                                                                                                                                                                                                                                           
  - Affected: 26.6.0 and 26.6.1 (both tested)
  - Not affected: 26.5.7 (works cleanly against the same DB state)                                                                                                                                                                                                                                  
  - Quarkus: 3.33.1 (affected) vs 3.27.3 (not affected) — large Quarkus bump between 26.5 and 26.6
                                                                                                                                                                                                                                                                             
**_Log excerpt:_**                                                                                                                                                                                                                                                                          
  ```
  14:51:27,476 DEBUG [QuarkusJpaConnectionProviderFactory] Calling migrateModel                                                                                                                                                                                                                     
  14:51:27,518 DEBUG [CustomLockService] Successfully acquired database lock                                                                                                                                                                                                                        
  14:51:28,000 INFO  [DefaultMigrationManager] Migrating older model to 26.6.1                                                                                                                                                                                                                      
  [... all 30 realms migrate, ~5 seconds ...]                                                                                                                                                                                                                                                       
  14:51:29,580 INFO  [RealmMigration] migrated realm <last-realm> to 26.6.1                                                                                                                                                                                                                         
  14:51:29,649 DEBUG [LiquibaseDBLockProvider] Going to release database lock namespace=DATABASE                                                                                                                                                                                                    
  14:51:29,684 DEBUG [io.quarkus.runtime.Application] Stopping application                                                                                                                                                                                                                          
  14:51:29,684 DEBUG [io.quarkus.runtime.shutdown.ShutdownRecorder] Attempting to gracefully shutdown.                                                                                                                                                                                              
  [Quarkus shuts down cleanly; exit 1] 
```                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                    
**_Workaround_**: pin back to quay.io/keycloak/keycloak:26.5.7. Works cleanly.

### Version

26.6.0

### Regression

- [x] The issue is a regression

### Expected behavior

KC should persist realm migrations to migration_model, continue startup through provider postInit(), and serve requests. It should not exit after migration.

### Actual behavior

KC exits cleanly (exit 1, shutdown hooks run) ~100ms after the last realm migrates and the Liquibase lock is released. The migration_model table shows no 26.6.x row, confirming the migration writes were rolled back. 

### How to Reproduce?

1. Start with a Postgres DB where migration_model has 26.5.6 as the latest version (i.e., any production DB that's been through the 26.5 series).                                                                                                                                                 
2. Deploy Keycloak 26.6.0 or 26.6.1 against that DB.                                                                                                                                                                                                                                              
3. Observe the container start, migrate, and exit. ECS/K8s will restart it; the loop will repeat indefinitely.

### Anything else?

**_Environment:_**                                                                                                                                                                                                                                                                                      
  - Keycloak 26.6.0 and 26.6.1 (base quay.io/keycloak/keycloak:26.6.x)                                                                                                                                                                                                                              
  - Custom image adds themes and 3 Java providers (issue reproduces with and without them)
  - AWS ECS on Bottlerocket, single replica in                                                                                                                                                                                                                                                 
  - PostgreSQL 13.20 on AWS RDS                   
  - ~30 realms, upgraded linearly through every 26.x version since 12.0.4                                                                                                                                                                                                                           
  - Log: databasechangelog shows schema-level Liquibase migrations for 26.6 committed successfully (the issue is exclusively at the realm-model migration step)
 
**_Ruled out through testing:_**                                                                                                                                                                                                                                                                        
  - Custom providers — removed all three custom provider jars from /opt/keycloak/providers/ before startup; same symptom.                                                                                                                                                                           
  - OpenTelemetry Java agent — disabled agent injection; same symptom.                                                                                                                                                                                                                              
  - JGroups mTLS / CertificateReloadManager (new in 26.6) — set KC_CACHE_EMBEDDED_MTLS_ENABLED=false; same symptom.                                                                                                                                                                                 
  - Memory/OOM — container has 2GB, heap sized to 1.5GB, no OOMError anywhere in logs.                             
  - Feature flags — removed admin-fine-grained-authz:v1; same symptom.                                                                                                                                                                                                                              
  - Auto-build path — pre-built the image with kc.sh build --db=postgres --features=token-exchange --transaction-xa-enabled=false --health-enabled=true and started with --optimized; the auto-build path was bypassed, but KC still exited after migration.
  - Healthcheck interference — ALB grace period is 180s, crash happens at ~17s, so ALB/ECS is not killing the container. The task's "Stopped reason" in AWS ECS is Essential container in task exited with exit code 1.

I wrote a ByteBuddy-based Java agent that instruments Quarkus + Keycloak exit paths and prints stack traces on entry:

- `ApplicationLifecycleManager.exit(int)` — catches programmatic `asyncExit`
- `ApplicationLifecycleManager$ShutdownHookThread.run()` — catches SIGTERM
- `ApplicationLifecycleManager.waitForExit()` — catches when the main   thread's blocking wait returns unexpectedly
- `KeycloakMain.run()` — catches entry/return of the KC-as-QuarkusApplication
- **`QuarkusKeycloakApplication.exit(Throwable)`** —  this method receives the real exception, silently discards it, and calls  `Quarkus.asyncExit(1)`. The agent prints the discarded `Throwable`.

With the agent attached, the next crash produced the full stack:
```
2026-04-24 13:16:11,190 INFO  [...RealmMigration] migrated realm testsher to 26.6.1
2026-04-24 13:16:11,190 INFO  [...RealmMigration] migrated realm sftpmssm to 26.6.1
[EXIT-TRACER] >>> QuarkusKeycloakApplication.exit(Throwable) — THIS IS THE SWALLOWED CAUSE thread=pool-8-thread-1
[EXIT-TRACER] cause class: java.util.concurrent.CompletionException
[EXIT-TRACER] cause message: java.lang.NullPointerException: Cannot invoke
    "org.keycloak.models.AuthenticationFlowModel.getId()" because "flow" is null

java.util.concurrent.CompletionException: java.lang.NullPointerException:
    Cannot invoke "org.keycloak.models.AuthenticationFlowModel.getId()"
    because "flow" is null
  at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(...)
  at java.base/java.util.concurrent.CompletableFuture.completeThrowable(...)
  at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(...)
  ...
Caused by: java.lang.NullPointerException: Cannot invoke
    "org.keycloak.models.AuthenticationFlowModel.getId()" because "flow" is null
  at org.keycloak.models.utils.DefaultAuthenticationFlows.addOrganizationBrowserFlowStep(DefaultAuthenticationFlows.java:829)
  at org.keycloak.migration.migrators.MigrateTo26_6_1.migrateRealm(MigrateTo26_6_1.java:21)
  at org.keycloak.migration.migrators.RealmMigration.lambda$migrate$0(RealmMigration.java:46)
  at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(...)
  ...
  at org.keycloak.migration.migrators.RealmMigration.migrate(RealmMigration.java:36)
  at org.keycloak.storage.datastore.DefaultMigrationManager.migrate(DefaultMigrationManager.java:168)
  at org.keycloak.migration.MigrationModelManager.migrate(MigrationModelManager.java:33)
  at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.migrateModel(QuarkusJpaConnectionProviderFactory.java:223)
  at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.initSchema(QuarkusJpaConnectionProviderFactory.java:214)
  ...
  at org.keycloak.services.resources.KeycloakApplication.runBootstrap(KeycloakApplication.java:114)
  at org.keycloak.services.resources.KeycloakApplication.lambda$startup$0(KeycloakApplication.java:94)
  at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(...)
  ... 3 more

[EXIT-TRACER] >>> ApplicationLifecycleManager.exit(1) thread=pool-8-thread-1 daemon=false
java.lang.RuntimeException: [EXIT-TRACER] exit(int) caller stack
  at io.quarkus.runtime.ApplicationLifecycleManager.exit(ApplicationLifecycleManager.java:395)
  at io.quarkus.runtime.Quarkus.asyncExit(Quarkus.java:163)
  at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.exit(QuarkusKeycloakApplication.java:66)
  at org.keycloak.services.resources.KeycloakApplication.lambda$startup$1(KeycloakApplication.java:96)
  at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(...)
  ...
```

pointing directly at
`MigrateTo26_6_1.migrateRealm` and `DefaultAuthenticationFlows.addOrganizationBrowserFlowStep`.

https://github.com/keycloak/keycloak/blob/main/model/storage-private/src/main/java/org/keycloak/migration/migrators/MigrateTo26_6_1.java
```java
public void migrateRealm(KeycloakSession session, RealmModel realm) {
    DefaultAuthenticationFlows.addOrganizationBrowserFlowStep(
        realm,
        realm.getFlowByAlias(DefaultAuthenticationFlows.BROWSER_FLOW)  // returns null for the realm demo
    );
}
```
https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/models/utils/DefaultAuthenticationFlows.java
```java
public static void addOrganizationBrowserFlowStep(RealmModel realm, AuthenticationFlowModel flow) {
    if (!Profile.isFeatureEnabled(Feature.ORGANIZATION)) {
        return;
    }
    if (isOrganizationAuthenticatorPresent(realm, flow.getId())) {  // NPE: flow is null
        return;
    }
    // ...
}
```

The migrator passes `realm.getFlowByAlias("browser")` straight into the helper with no null check. For any realm missing a flow aliased `browser`, this cause nullPointerError

I run against the Keycloak DB this SQL
```sql
SELECT r.id, r.name
FROM realm r
LEFT JOIN authentication_flow af
  ON af.realm_id = r.id AND af.alias = 'browser'
WHERE af.id IS NULL;

                  id                  |   name
--------------------------------------+----------
 6a7abf3f-c444-4cca-b508-26cbe141e7c9 | demo
```
To fix this, since this is just a demo realm, I delete the whole realm and the upgrade is successful.

@vramik you were right, the exception was swallowed during async bootstrap, on setting KC_SERVER_ASYNC_BOOTSTRAP=false: wish I'd known before I wrote the bytebuddy agent 😅
So I think we should add a null-check at the call site MigrateTo26_6_1.migrateRealm

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java
@@ -18,8 +18,10 @@
 package org.keycloak.quarkus.runtime;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.Properties;
 import java.util.concurrent.ForkJoinPool;
+import java.util.function.Consumer;
 
 import jakarta.enterprise.context.ApplicationScoped;
 
@@ -57,6 +59,7 @@ import static org.keycloak.quarkus.runtime.Environment.hasEarlyExitLaunchMode;
 public class KeycloakMain implements QuarkusApplication {
 
     private static AbstractNonServerCommand COMMAND;
+    private static Consumer<Throwable> ERROR_HANDLER;
 
     static {
         InfinispanUtils.configureVirtualThreads();
@@ -132,6 +135,9 @@ public class KeycloakMain implements QuarkusApplication {
 
     public static void start(Picocli picocli, AbstractNonServerCommand command, ExecutionExceptionHandler errorHandler) {
         COMMAND = command; // it would be nice to not do this statically - start quarkus with an instance of KeycloakMain, rather than a class for example
+        ERROR_HANDLER = cause -> errorHandler.error(picocli.getErrWriter(),
+                String.format("Failed to start server in (%s) mode", getKeycloakModeFromProfile(org.keycloak.common.util.Environment.getProfile())),
+                cause.getCause());
         try {
             Quarkus.run(KeycloakMain.class, (exitCode, cause) -> {
                 if (cause != null) {
@@ -145,6 +151,9 @@ public class KeycloakMain implements QuarkusApplication {
             errorHandler.error(picocli.getErrWriter(),
                     String.format("Unexpected error when starting the server in (%s) mode", getKeycloakModeFromProfile(org.keycloak.common.util.Environment.getProfile())),
                     cause.getCause());
+        } finally {
+            ERROR_HANDLER = null;
+            COMMAND = null;
         }
         picocli.exit(CommandLine.ExitCode.SOFTWARE);
     }
@@ -169,5 +178,10 @@ public class KeycloakMain implements QuarkusApplication {
 
         return ApplicationLifecycleManager.getExitCode();
     }
-
+    
+    public static void asyncExit(int exitCode, Throwable t) {
+        Optional.ofNullable(ERROR_HANDLER).ifPresent(h -> h.accept(t));
+        Quarkus.asyncExit(exitCode);
+    }
+    
 }

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/jaxrs/QuarkusKeycloakApplication.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/jaxrs/QuarkusKeycloakApplication.java
@@ -17,6 +17,8 @@
 
 package org.keycloak.quarkus.runtime.integration.jaxrs;
 
+import java.util.concurrent.CompletableFuture;
+
 import jakarta.enterprise.event.Observes;
 import jakarta.ws.rs.ApplicationPath;
 
@@ -25,6 +27,7 @@ import org.keycloak.config.ServerOptions;
 import org.keycloak.connections.jpa.JpaConnectionProvider;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.quarkus.runtime.Environment;
+import org.keycloak.quarkus.runtime.KeycloakMain;
 import org.keycloak.quarkus.runtime.configuration.Configuration;
 import org.keycloak.quarkus.runtime.configuration.MicroProfileConfigProvider;
 import org.keycloak.quarkus.runtime.configuration.PropertyMappingInterceptor;
@@ -37,11 +40,11 @@ import org.keycloak.services.resources.KeycloakApplication;
 import org.keycloak.utils.StringUtil;
 
 import io.quarkus.arc.Arc;
-import io.quarkus.runtime.Quarkus;
 import io.quarkus.runtime.ShutdownDelayInitiatedEvent;
 import io.quarkus.runtime.ShutdownEvent;
 import io.quarkus.runtime.StartupEvent;
 import io.smallrye.common.annotation.Blocking;
+import org.eclipse.microprofile.context.ManagedExecutor;
 import org.jboss.logging.Logger;
 
 import static org.keycloak.common.util.Environment.isDevMode;
@@ -65,13 +68,20 @@ public class QuarkusKeycloakApplication extends KeycloakApplication {
         });
     }
 
-    @Override
-    protected void exit(Throwable cause) {
-        Quarkus.asyncExit(1);
-    }
-
     void onStartupEvent(@Observes StartupEvent event) {
-        startup();
+        var asyncBootstrap = Configuration.getOptionalKcValue(ServerOptions.SERVER_ASYNC_BOOTSTRAP)
+                .map(Boolean::parseBoolean)
+                .orElse(Boolean.TRUE);
+        // skip async bootstrap in dev and non-server mode
+        if (isDevMode() || isNonServerMode() || hasEarlyExitLaunchMode() || !asyncBootstrap) {
+            startup();      
+        } else {
+            ManagedExecutor executor = Arc.container().instance(ManagedExecutor.class).get();
+            CompletableFuture.runAsync(this::startup, executor).exceptionally(cause -> {
+                KeycloakMain.asyncExit(1, cause);
+                return null;
+            });        
+        }
     }
 
     void onShutdownEvent(@Observes ShutdownEvent event) {
@@ -85,11 +95,6 @@ public class QuarkusKeycloakApplication extends KeycloakApplication {
     @Override
     public DefaultKeycloakSessionFactory createSessionFactory() {
         return Arc.container().instance(QuarkusKeycloakSessionFactory.class).get();
-    }
-
-    @Override
-    protected void initAndStart() {
-        // no need - is handled by Quarkus logic and onStartupEvent
     }
 
     @Override
@@ -111,15 +116,6 @@ public class QuarkusKeycloakApplication extends KeycloakApplication {
         } catch (NumberFormatException e) {
             throw new RuntimeException("Invalid admin expiration value provided. An integer is expected.", e);
         }
-    }
-
-    @Override
-    protected boolean supportsAsyncInitialization() {
-        var asyncBootstrap = Configuration.getOptionalKcValue(ServerOptions.SERVER_ASYNC_BOOTSTRAP)
-                .map(Boolean::parseBoolean)
-                .orElse(Boolean.TRUE);
-        // skip async bootstrap in dev and non-server mode
-        return !isDevMode() && !isNonServerMode() && !hasEarlyExitLaunchMode() && asyncBootstrap;
     }
 
     @Override

--- a/quarkus/tests/integration/src/test-providers/java/org/keycloak/it/resource/realm/TestRealmResourceFactory.java
+++ b/quarkus/tests/integration/src/test-providers/java/org/keycloak/it/resource/realm/TestRealmResourceFactory.java
@@ -28,6 +28,7 @@ import org.keycloak.services.resource.RealmResourceProviderFactory;
  */
 public class TestRealmResourceFactory implements RealmResourceProviderFactory {
     public static final String ID = "test-resources";
+    private boolean fail;
 
     @Override
     public RealmResourceProvider create(KeycloakSession session) {
@@ -36,12 +37,14 @@ public class TestRealmResourceFactory implements RealmResourceProviderFactory {
 
     @Override
     public void init(Config.Scope config) {
-
+        fail = Boolean.parseBoolean(config.get("fail", null));
     }
 
     @Override
     public void postInit(KeycloakSessionFactory factory) {
-
+        if (fail) {
+            throw new RuntimeException("I've failed");
+        }
     }
 
     @Override

--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/RawKeycloakDistribution.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/RawKeycloakDistribution.java
@@ -477,7 +477,7 @@ public final class RawKeycloakDistribution implements KeycloakDistribution {
         return socketFactory;
     }
 
-    private boolean isRunning() {
+    public boolean isRunning() {
         return keycloak != null && keycloak.isAlive();
     }
 

--- a/services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java
+++ b/services/src/main/java/org/keycloak/services/resources/KeycloakApplication.java
@@ -19,8 +19,6 @@ package org.keycloak.services.resources;
 import java.io.File;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 
 import jakarta.ws.rs.core.Application;
@@ -54,16 +52,6 @@ public abstract class KeycloakApplication extends Application {
 
     private static volatile DefaultKeycloakSessionFactory sessionFactory;
 
-    public KeycloakApplication() {
-        try {
-            initTmpDirectory();
-            logger.debugv("Application: {0}", this.getClass().getName());
-            initAndStart();
-        } catch (Throwable t) {
-            exit(t);
-        }
-    }
-
     public static String getTmpDirectory() {
         return System.getProperty(KC_TMPDIR, System.getProperty("java.io.tmpdir"));
     }
@@ -80,34 +68,19 @@ public abstract class KeycloakApplication extends Application {
         System.setProperty(KC_TMPDIR, tmpDir.getAbsolutePath());
     }
 
-    protected abstract void exit(Throwable t);
-
     protected abstract String getDataDir();
 
-    protected void startup() {
+    // synchronized to prevent shutdown while running bootstrapping
+    protected synchronized void startup() {
+        logger.debugv("Application: {0}", this.getClass().getName());
+        initTmpDirectory();
         Profile.getInstance().logUnsupportedFeatures();
         CryptoIntegration.init(KeycloakApplication.class.getClassLoader());
         KeycloakApplication.sessionFactory = createSessionFactory();
-
-        if (supportsAsyncInitialization()) {
-            final var executor = Executors.newSingleThreadExecutor();
-            CompletableFuture.runAsync(() -> runBootstrap(KeycloakApplication.sessionFactory), executor)
-                    .exceptionally(throwable -> {
-                        exit(throwable);
-                        return null;
-                    })
-                    .thenRun(executor::shutdown);
-            return;
-        }
-
         runBootstrap(KeycloakApplication.sessionFactory);
     }
 
-    protected boolean supportsAsyncInitialization() {
-        return false;
-    }
-
-    private synchronized void runBootstrap(DefaultKeycloakSessionFactory keycloakSessionFactory) {
+    private void runBootstrap(DefaultKeycloakSessionFactory keycloakSessionFactory) {
         var startTime = System.nanoTime();
 
         keycloakSessionFactory.init();
@@ -189,10 +162,12 @@ public abstract class KeycloakApplication extends Application {
 
     protected abstract void createTemporaryAdmin(KeycloakSession session);
 
-    protected abstract void initAndStart();
-
     protected abstract DefaultKeycloakSessionFactory createSessionFactory();
 
+    /**
+     * WARNING: This method is for use by test logic. Will return null if there is no current KeycloakApplication, or if the
+     * startup has not yet reached the point of setting this value.
+     */
     public static DefaultKeycloakSessionFactory getSessionFactory() {
         return sessionFactory;
     }

--- a/services/src/test/java/org/keycloak/services/resteasy/ResteasyKeycloakApplication.java
+++ b/services/src/test/java/org/keycloak/services/resteasy/ResteasyKeycloakApplication.java
@@ -68,16 +68,18 @@ public class ResteasyKeycloakApplication extends KeycloakApplication {
             // an endpoint for the load balancer to gather information whether this site should receive requests or not.
             classes.add(LoadBalancerResource.class);
         }
+        Config.init(new JsonConfigProviderFactory().create()
+                .orElseThrow(() -> new RuntimeException("Failed to load Keycloak configuration")));
+        Profile.configure(
+                new PropertiesProfileConfigResolver(System.getProperties()),
+                new PropertiesFileProfileConfigResolver()
+        );
+        startup();
     }
 
     @Override
     protected String getDataDir() {
         return System.getProperty("project.build.directory");
-    }
-
-    @Override
-    protected void exit(Throwable cause) {
-        throw new RuntimeException(cause);
     }
 
     @Override
@@ -98,17 +100,6 @@ public class ResteasyKeycloakApplication extends KeycloakApplication {
     @Override
     protected void createTemporaryAdmin(KeycloakSession session) {
         // do nothing
-    }
-
-    @Override
-    protected void initAndStart() {
-        Config.init(new JsonConfigProviderFactory().create()
-                .orElseThrow(() -> new RuntimeException("Failed to load Keycloak configuration")));
-        Profile.configure(
-                new PropertiesProfileConfigResolver(System.getProperties()),
-                new PropertiesFileProfileConfigResolver()
-        );
-        startup();
     }
 
     @Override

